bd-log-primitives/src/lib.rs

```diff
@@ -215,38 +252,239 @@ pub enum LogFieldKind {
+    os.write_raw_varint32(Self::size_proto_data(value).to_u32())?;
+    match value {
+      StringOrBytes::String(s) => {
+        os.write_string(1, s)?;
+      },
+      StringOrBytes::SharedString(s) => {
+        os.write_string(1, s.as_ref())?;
+      },
+      StringOrBytes::Bytes(b) => {
+        // This encodes the Binary proto message.
+        let inner_len = ::protobuf::rt::bytes_size(2, b);
+        os.write_tag(2, WireType::LengthDelimited)?;
+        os.write_raw_varint32(inner_len.to_u32())?;
+        os.write_bytes(2, b)?;
+      },
+    }
+
+    Ok(())
+  }
+
+  fn size_proto_field(key: &LogFieldKey, value: &LogFieldValue) -> u64 {
+    let mut my_size = 0;
+    my_size += ::protobuf::rt::string_size(1, key.as_ref());
+    let value_len = Self::size_proto_data(value);
+    my_size += 1 + ::protobuf::rt::compute_raw_varint64_size(value_len) + value_len;
+    my_size
+  }
+
+  fn serialize_proto_field(
+    field_number: u32,
+    key: &LogFieldKey,
+    value: &LogFieldValue,
+    os: &mut CodedOutputStream<'_>,
+  ) -> anyhow::Result<()> {
+    os.write_tag(field_number, WireType::LengthDelimited)?;
+    os.write_raw_varint32(Self::size_proto_field(key, value).to_u32())?;
+    os.write_string(1, key.as_ref())?;
+    Self::serialize_proto_data(2, value, os)?;
+    Ok(())
+  }
+
+  #[must_use]
+  pub fn serialize_proto_size_inner(
+    log_level: u32,
+    message: &LogFieldValue,
+    fields: &LogFields,
+    session_id: &str,
+    occurred_at: OffsetDateTime,
+    log_type: LogType,
+    action_ids: &[&str],
+    stream_ids: &[&str],
+  ) -> u64 {
+    let mut my_size = 0;
+
+    my_size += ::protobuf::rt::uint64_size(1, occurred_at.unix_timestamp_nanos().to_u64() / 1000);
 
-  pub capture_session: Option<&'a str>,
+    if log_level != 0 {
+      my_size += ::protobuf::rt::uint32_size(2, log_level);
+    }
+
+    let message_len = Self::size_proto_data(message);
+    my_size += 1 + ::protobuf::rt::compute_raw_varint64_size(message_len) + message_len;
+
+    for value in fields {
+      let len = Self::size_proto_field(value.0, value.1);
+      my_size += 1 + ::protobuf::rt::compute_raw_varint64_size(len) + len;
+    }
+
+    my_size += ::protobuf::rt::string_size(5, session_id);
+
+    for value in action_ids {
+      my_size += ::protobuf::rt::string_size(6, value);
+    }
+
+    if log_type != LogType::NORMAL {
+      my_size += ::protobuf::rt::int32_size(7, log_type as i32);
+    }
+
+    for value in stream_ids {
+      my_size += ::protobuf::rt::string_size(8, value);
+    }
+
+    my_size
+  }
+
+  #[must_use]
+  pub fn serialized_proto_size(&self, action_ids: &[&str], stream_ids: &[&str]) -> u64 {
+    Self::serialize_proto_size_inner(
+      self.log_level,
+      &self.message,
+      &self.fields,
+      &self.session_id,
+      self.occurred_at,
+      self.log_type,
+      action_ids,
+      stream_ids,
+    )
+  }
+
+  pub fn serialize_proto_to_stream_inner(
```

Comment:
Obviously this means that we need to maintain this as more fields are added - any way to have a test fail in that case?

Reply:
Yeah can do.